### ISSUE TITLE
Check test coverage for Rust code on CI

### ISSUE THREAD

**Issue by Minoru** [newsboat/newsboat#346]
> It's time for me to figure out how to check test coverage for Rust code. With C++, these "non-tests" would've triggered a warning from Coveralls.

_Originally posted by @Minoru in https://github.com/newsboat/newsboat/pull/342_

**Comment by Minoru** [comment]
As far as I can see, there are three options for Rust at the moment:
- `kcov`, usually wrapped in [`cargo-travis`](https://github.com/roblabla/cargo-travis);
- [`tarpaulin`](https://github.com/xd009642/tarpaulin)
- [`grcov`](https://github.com/mozilla/grcov)

But:
- all three don't support anything but Linux x86_64
- `grcov` documentation is very spotty, [people have trouble running it](https://users.rust-lang.org/t/test-coverage-with-grcov/24851)
- `grcov` requires Nightly Rust and yet-unstabilized [`-Z profile`](https://github.com/rust-lang/rust/issues/42524)
- `kcov` gained Rust support relatively recently (version 31, it seems). It's absent from Ubuntu Bionic, so on Travis we'd have to built it from source. On Cirrus we can probably use one of the [pre-built containers](https://github.com/Ragnaroek/kcov_docker)
- rustc arranges for the linker to strip unused code, we [need to combat that](https://stackoverflow.com/a/38371687/2350060) in order to get accurate stats from `kcov`
- `tarpaulin` is still marked as experimental, but seem the most workable right now—or maybe no-one uses it because I haven't seen many complaints
- `kcov` can't calculate coverage for doctests

Also:
- `tarpaulin` integrates with Cargo, so we won't be able to calculate C++ coverage with it (I guess)
- `grcov` relies on LLVM's instrumentation, so I assume we'd be able to get C++ coverage if we pass some options to `clang++`—but GCC builds won't be able to submit coverage reports
- `kcov` is language-agnostic

I'm a bit beaten down by all this, so I'm laying this issue aside for now.

(This was originally posted to another issue, I'm moving it here now).

**Comment by Minoru** [comment]
I tried `tarpaulin` and `grcov` locally today.

`tarpaulin` is easy as pie: `cargo install cargo-tarpaulin`, and then just `cargo tarpaulin` to generate the coverage. It runs tests sequentially and slowly, so takes a while. Also, it doesn't include C++ coverage, just as I feared.

`grcov` indeed doesn't work with GCC, but with Clang, it's all smooth: `grcov . -o lcov.info` produces a file that `genhtml` can understand, and it appears we can make it produce Coveralls-compatible output with `-t coveralls`. That's what I'm going to try next.

(Given how well these two work, I'm leaving `kcov` out of consideration for now—lack of doctests support is a big drawback IMHO, and it appears that `kcov` won't be able to make up for it given how good the competition is.)
